formats/json/commonTest/src/kotlinx/serialization/JsonPathTest.kt

```diff
@@ -0,0 +1,162 @@
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.json.*
+import kotlinx.serialization.test.*
+import kotlin.test.*
+
+class JsonPathTest : JsonTestBase() {
+
+    @Serializable
+    class Outer(val a: Int, val i: Inner)
+
+    @Serializable
+    class Inner(val a: Int, val b: String, val c: List<String>, val d: Map<Int, Box>)
+
+    @Serializable
+    class Box(val s: String)
+
+    @Test
+    fun testBasicError() {
+        expectPath("$.a") { Json.decodeFromString<Outer>("""{"a":foo}""") }
+        expectPath("$.i") { Json.decodeFromString<Outer>("""{"a":42, "i":[]}""") }
+        expectPath("$.i.b") { Json.decodeFromString<Outer>("""{"a":42, "i":{"a":43, "b":42}""") }
+        expectPath("$.i.b") { Json.decodeFromString<Outer>("""{"a":42, "i":{"b":42}""") }
+    }
+
+    @Test
+    fun testMissingKey() {
+        expectPath("$.i.d[1]") { Json.decodeFromString<Outer>("""{"a":42, "i":{"d":{1:{}}""") }
+    }
+
+    @Test
+    fun testUnknownKeyIsProperlyReported() {
+        expectPath("$.i") { Json.decodeFromString<Outer>("""{"a":42, "i":{"foo":42}""") }
+        expectPath("$") { Json.decodeFromString<Outer>("""{"x":{}, "a": 42}""") }
+//        Json.decodeFromString<Outer>("""{"a":42, "x":{}}""")
+    }
+
+    @Test
+    fun testMalformedRootObject() {
+        expectPath("$") { Json.decodeFromString<Outer>("""{{""") }
+    }
+
+    @Test
+    fun testArrayIndex() {
+        expectPath("$.i.c[1]") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "c": ["a", 2]}""") }
+        expectPath("$[2]") { Json.decodeFromString<List<String>>("""["a", "2", 3]""") }
+    }
+
+    @Test
+    fun testArrayIndexMalformedArray() {
+        // Also zeroes as we cannot distinguish what exactly wen wrong is such cases
+        expectPath("$.i.c[0]") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "c": [[""") }
+        expectPath("$[0]") { Json.decodeFromString<List<String>>("""[[""") }
+        // But we can here
+        expectPath("$.i.c\n") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "c": {}}}""") }
+        expectPath("$\n") { Json.decodeFromString<List<String>>("""{""") }
+    }
+
+    @Test
+    fun testMapKey() {
+        expectPath("$.i.d\n") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "d": {"foo": {}}""") }
+        expectPath("$.i.d\n") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "d": {42: {"s":"s"}, 42.0:{}}""") }
+        expectPath("$\n") { Json.decodeFromString<Map<Int, String>>("""{"foo":"bar"}""") }
+        expectPath("$\n") { Json.decodeFromString<Map<Int, String>>("""{42:"bar", "foo":"bar"}""") }
+    }
+
+    @Test
+    fun testMalformedMap() {
+        expectPath("$.i.d\n") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "d": []""") }
+        expectPath("$\n") { Json.decodeFromString<Map<Int, String>>("""[]""") }
+    }
+
+    @Test
+    fun testMapValue() {
+        expectPath("$.i.d[42]\n") { Json.decodeFromString<Outer>("""{"a":42, "i":{ "d": {42: {"xx":"bar"}}""") }
```

Comment:
Shouldn't these be quoted such as `$.i.d['42']` to indicate object keys and not array indexing? As of now it is indistinguishable (and thus ambiguous) from the 43rd element of an array.